build_tools/travis/flake8_diff.sh

```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+
+set -e
+
+PROJECT=scikit-learn/scikit-learn
+PROJECT_URL=https://github.com/$PROJECT.git
+
+echo "Remotes:"
+git remote --verbose
+
+# Find the remote with the project name (upstream in most cases)
+REMOTE=$(git remote -v | grep $PROJECT | cut -f1 | head -1)
+
+# Add a temporary remote if needed. For example this is necessary when
+# Travis is configured to run in a fork. In this case 'origin' is the
+# fork and not the reference repo we want to diff against.
+if [[ -z "$REMOTE" ]]; then
+    TMP_REMOTE=tmp_reference_upstream
+    REMOTE=$TMP_REMOTE
+    git remote add $REMOTE $PROJECT_URL
+fi
+
+if [[ "$TRAVIS" == "true" ]]; then
+    if [[ "$TRAVIS_PULL_REQUEST" == "false" ]]
+    then
+        # Travis does the git clone with a limited depth (50 at the time of
+        # writing). This may not be enough to find the common ancestor with
+        # $REMOTE/master so we unshallow the git checkout
+        git fetch --unshallow || echo "Unshallowing the git checkout failed"
+    else
+        # We want to fetch the code as it is in the PR branch and not
+        # the result of the merge into master. This way line numbers
+        # reported by Travis will match with the local code.
+        BRANCH_NAME=travis_pr_$TRAVIS_PULL_REQUEST
+        git fetch $REMOTE pull/$TRAVIS_PULL_REQUEST/head:$BRANCH_NAME
+        git checkout $BRANCH_NAME
```

Comment:
In my mind these lines were supposed of doing the same as what Travis does by default, but with the branch as it is on the PR author's machine instead of at is is after the merge into master. This way the line numbers on Travis and on the PR author machine match.

> I assume there are no concurrency issues with modifying the working copy in the current directory? The rest of this script only modifies .git. And I don't actually see why a working copy is needed.

Note this only happens on Travis (not if you run the command locally) so I am not sure about which concurrency issue you are talking about. It may be true that we may not need a working copy but again this happens only on Travis so is it such a  big problem ?